[Discover] Inline shard failures warnings

jughosta opened the pull request:
- Closes https://github.com/elastic/kibana/issues/155216

## Summary

This PR replaces shard falures toasts with inline warnings in Discover.

- [x] Intercept shard failures in Discover main app
- [x] Show inline warnings above the grid instead
- [x] Handle NoResultsFound case too
- [x] Implement for Discover context app
- [x] Implement for saved search embeddable on Dashboard
- [x] Can we inline timeouts too?
- [x] Check SQL view
- [x] Add tests


Discover view with shard failures
<img width="400" alt="Screenshot 2023-07-06 at 14 23 48" src="https://github.com/elastic/kibana/assets/1415710/a0799aa0-9d2e-42ee-b89b-e0e1180220a5">

Discover view with shard failures (and no results)
<img width="400" alt="Screenshot 2023-07-07 at 13 24 50" src="https://github.com/elastic/kibana/assets/1415710/28dc2bad-9776-4aa9-8f51-219d1c87487a">

Dashboard view with shard failures
<img width="400" alt="Screenshot 2023-07-06 at 16 15 49" src="https://github.com/elastic/kibana/assets/1415710/2a68c19a-1ca8-4f10-a9f5-4aa56f9160b0">

Surrounding documents view with shard failures
<img width="400" alt="Screenshot 2023-07-10 at 17 26 31" src="https://github.com/elastic/kibana/assets/1415710/ade63cfe-a1c2-4c22-8823-58dcfef9357f">

Discover view with timeouts
<img width="400" alt="Screenshot 2023-07-07 at 16 47 27" src="https://github.com/elastic/kibana/assets/1415710/0101be19-a555-4f96-b963-7fe418d51fb5">

Dashboard view with timeouts
<img width="400" alt="Screenshot 2023-07-07 at 16 48 18" src="https://github.com/elastic/kibana/assets/1415710/82979365-4129-4385-8a13-9c139e1acbf1">

Surrounding documents view with timeouts
<img width="400" alt="Screenshot 2023-07-11 at 15 03 41" src="https://github.com/elastic/kibana/assets/1415710/0ea41b79-ac6b-4456-9cfa-0d038b10da7d">

## Testing

For testing please uncomment https://github.com/elastic/kibana/blob/3f102cf688dbfe87fd901cf7f8b59680ff62e5b8/src/plugins/data/common/search/search_source/search_source.ts#L922 or https://github.com/elastic/kibana/blob/3f102cf688dbfe87fd901cf7f8b59680ff62e5b8/src/plugins/data/common/search/search_source/search_source.ts#L547 and switch to `kibana*` data view.


### Checklist

Delete any items that are not applicable to this PR.

- [x] Any text added follows [EUI's writing guidelines](https://elastic.github.io/eui/#/guidelines/writing), uses sentence case text and includes [i18n support](https://github.com/elastic/kibana/blob/main/packages/kbn-i18n/README.md)
- [x] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios
- [x] This renders correctly on smaller devices using a responsive layout. (You can test this [in your browser](https://www.browserstack.com/guide/responsive-testing-on-local-server))
- [x] This was checked for [cross-browser compatibility](https://www.elastic.co/support/matrix#matrix_browsers)

elasticmachine commented:
Pinging @elastic/kibana-data-discovery (Team:DataDiscovery)

davismcphee reviewed (commented):
Finished my initial code review and it's looking good so far! I left a couple of minor nits in the comments (feel free to ignore if preferred, nothing blocking), but mainly just had a couple of questions. I'm working on local testing now and will follow up with another review based on my findings there.

davismcphee reviewed (commented):
Pulled and tested and it's working great overall. I didn't find any bugs in the functionality, just a couple of UI issues I noticed and some questions.

For some reason the Discover cell action styles seem to be broken in this branch. I'm not sure what would have caused this, but it seems to be fine in `main`:
![cell_actions](https://github.com/elastic/kibana/assets/25592674/a220a364-04a8-47dc-b1d7-2dfb391980f3)

The spacing between the no data icon and title seemed to be a lot larger than in your screenshot:
<img width="1274" alt="no_data" src="https://github.com/elastic/kibana/assets/25592674/15642727-dbda-4f31-94c4-756a5654b486">

The way the text in the inline callout wraps on smaller screen sizes looks a bit crammed. I wonder if it would be worth truncating like we do in the inline error callout or if it's best to preserve the full message:
<img width="781" alt="word_wrap" src="https://github.com/elastic/kibana/assets/25592674/c7b9b62e-f53d-4afe-90b4-e8b358efb737">

Maybe less of a concern since we'd need to have a lot of warnings, but it could be worth it to add scrolling to the inline callout mode like the other modes to keep it from taking up too much space:
<img width="1287" alt="discover" src="https://github.com/elastic/kibana/assets/25592674/ea738cb4-af77-4764-802b-234862bbcdeb">

Also I really like the idea of using a badge similar to the Lens badge for the embeddable. My only concern is that it partially obscures the pagination, but not sure if that's really a concern. Maybe something to get some design feedback on.

jughosta commented on src/plugins/discover/public/utils/get_search_response_intercepted_warnings.tsx:
```diff
@@ -0,0 +1,71 @@
+
+  const interceptedWarnings: SearchResponseInterceptedWarning[] = [];
+
+  services.data.search.showWarnings?.(adapter, (warning, meta) => {
```
Let's see if some jest tests break

jughosta commented on src/plugins/discover/public/application/main/components/layout/discover_documents.tsx:
```diff
@@ -203,6 +204,13 @@ function DiscoverDocumentsComponent({
         </h2>
       </EuiScreenReaderOnly>
+      {!!documentState.interceptedWarnings?.length && (
+        <WarningsCallout
```
@davismcphee I think these warnings are relevant only for grid data. Field stats should handle its own data fetching.

davismcphee replied:
That's fair, I suppose this does only handle the grid requests. Let's leave it as is for now then. Maybe an opportunity for a future enhancement to do similar for the field stats.

jughosta commented on src/plugins/discover/public/components/common/warnings_callout/warnings_callout.scss:
Moved to emotion

jughosta commented:
@davismcphee Thanks for the review! I think I addressed your comments.

> For some reason the Discover cell action styles seem to be broken in this branch. I'm not sure what would have caused this, but it seems to be fine in main

It was fixed in main via https://github.com/elastic/kibana/pull/161592 and I merged it in today.

davismcphee reviewed (commented):
Thanks for the latest changes! I think the updated text wrapping is definitely an improvement over before, although for some reason I'm now getting some additional whitespace below the callout:
<img width="770" alt="whitespace" src="https://github.com/elastic/kibana/assets/25592674/d2df2097-59ae-4e49-9ae2-66af10eb5802">

Otherwise this looks good to me once the whitespace is figured out. I'm also going to request a review from @andreadelrio to get some design feedback on this.

jughosta commented:
@davismcphee Oops, broke it with the scroll. Should be better now.

jughosta commented:
@davismcphee @andreadelrio I implemented `close` buttons for each callout but now I am thinking of replacing it with just a single button which would close all warnings at once. Wdyt?

<img width="400" alt="Screenshot 2023-07-18 at 10 09 37" src="https://github.com/elastic/kibana/assets/1415710/900ceab4-7fb4-4894-a9b6-333f92e7f1c0">

drewdaemon reviewed (approved):
Search source comments LGTM

davismcphee reviewed (commented):
Looks like the whitespace issue is fixed, thanks! Unfortunately I noticed an issue I missed before. It looks like the toast still appears if the chart is hidden, presumably related to Unified Histogram:
<img width="1172" alt="chart_hidden" src="https://github.com/elastic/kibana/assets/25592674/9dafad19-ae60-4440-8920-ad6ecaaab58d">

>@davismcphee @andreadelrio I implemented close buttons for each callout but now I am thinking of replacing it with just a single button which would close all warnings at once. Wdyt?

That could be convenient! I'll leave it to @andreadelrio to make that call, but either approach would be acceptable IMO.

jughosta commented:
> It looks like the toast still appears if the chart is hidden, presumably related to Unified Histogram

@davismcphee Thanks for catching it! How would you suggest to handle it? I see several options: 
- keep the toast for this case
- suppress the toast without any feedback as it will most probably be covered by warnings above the grid
- suppress the toast and add a warning icon next to the hits number too.

jughosta commented:
Hi @davismcphee,

I rearranged the components and moved the code to a new package `@kbn/search-response-warnings` so we can use it in unifiedHistogram for a warning badge for example. I think I also addressed the bug and now hits number is shown when histogram is hidden and there are some shard failures.

I have not added a warning badge to the histogram hits yet, only suppressed the toast. I think I would need your help to understand how discover `totalHits$` interacts the histogram hits state and when the fetching is happening.

davismcphee reviewed (approved):
Thanks for the latest changes! The `@kbn/search-response-warnings` package looks good to me, and I confirmed locally that Unified Histogram is now correctly displaying the hits count on shard failures 👍

I created a separate issue #162604 to track implementing the badge in Unified Histogram since as you mentioned, it's more effort due to how `totalHits$` works and the partial results handling. I think we should be good to merge these changes in the meantime and work on  #162604 as a followup once we've had time to look into it more.

kertal commented:
Adding a note, because taking care of totalHits$ is also related to this issue: [#155061](https://github.com/elastic/kibana/issues/155061)

bhavyarm reviewed (approved):
LGTM. Pulled the PR and did a quick run through changes. Also yay awesome tests.

andreadelrio reviewed (changes requested):
I was unable to get all the examples showing up locally so sending feedback mostly based on screenshots. See feedback below.

> Discover view with shard failures 
<img alt="Screenshot 2023-07-06 at 14 23 48" width="400" src="https://user-images.githubusercontent.com/1415710/251448809-a0799aa0-9d2e-42ee-b89b-e0e1180220a5.png">

> Discover view with shard failures (and no results) 

Remove gray border this empty prompt
<img alt="Screenshot 2023-07-07 at 13 24 50" width="400" src="https://user-images.githubusercontent.com/1415710/251729570-28dc2bad-9776-4aa9-8f51-219d1c87487a.png">
> 
> Dashboard view with shard failures

The button showing the error shouldn't be overlapping the panels' content. We should show it below in a separate row
<img alt="Screenshot 2023-07-06 at 16 15 49" width="400" src="https://user-images.githubusercontent.com/1415710/251478726-2a68c19a-1ca8-4f10-a9f5-4aa56f9160b0.png">
<img width="657" alt="Frame 3" src="https://github.com/elastic/kibana/assets/4016496/043dfdbf-d8f6-4e09-8d9f-eb7c8396b8ff">

> Surrounding documents view with shard failures 
<img width="1440" alt="Frame 4" src="https://github.com/elastic/kibana/assets/4016496/0e08b006-bc06-4dd9-a1db-11812f209188">

- Question, how did we make the decision to show these as warning (yellow) rather than errors (red)?

jughosta commented:
Thanks @andreadelrio!

### 1
> The button showing the error shouldn't be overlapping the panels' content. We should show it below in a separate row

Updated the positioning of the badge:
<img width="400" alt="Screenshot 2023-08-02 at 11 22 06" src="https://github.com/elastic/kibana/assets/1415710/d9385ed7-00e0-462c-b09b-e0fdcc769943">
<img width="400" alt="Screenshot 2023-08-02 at 11 21 44" src="https://github.com/elastic/kibana/assets/1415710/36ab3540-4af2-4527-a415-616a7fd2907c">


### 2
> Remove gray border this empty prompt

You mean to make it look like the following? Might be not enough contrast. Is there anything else we can use from eui for this case? It's when `0` hits returned and there were some shards failures.

<img width="400" alt="Screenshot 2023-08-02 at 11 41 08" src="https://github.com/elastic/kibana/assets/1415710/56a284b0-7b6f-48fa-81e2-bbb7b88f0c55">


### 3
> Can we group duplicated messages?

Yes, they are grouped based on title and warnings content. The ones from the screenshot have the same title but different details inside.

### 4
> Why the bolt icon? what is it trying to convey and how does it related to the message?

This part I didn't touch in the PR. I guess the idea was to bring users attention that they reached the earliest or the latest document and there is nothing else to load. Would you suggest to change this callout?

kertal commented:
> ### 4
> > Why the bolt icon? what is it trying to convey and how does it related to the message?
> 
> This part I didn't touch in the PR. I guess the idea was to bring users attention to the fact that they reached the earliest or the latest document and there is nothing else to load. Would you suggest to change this callout?

That's legacy. There could be an info icon + the matching color for that callout ... because I don't think this needs to be a kind of warning.

andreadelrio commented:
Thanks @jughosta for making the updates.

### 2
> You mean to make it look like the following? Might be not enough contrast. Is there anything else we can use from eui for this case? It's when `0` hits returned and there were some shards failures.

Yes. I don't think the border does a lot to improve contrast tbh and no-border is the recommended pattern in EUI. See https://elastic.github.io/eui/#/display/empty-prompt/guidelines

<img width="400" alt="Screenshot 2023-08-02 at 11 41 08" src="https://github.com/elastic/kibana/assets/1415710/56a284b0-7b6f-48fa-81e2-bbb7b88f0c55">


### 3
> Yes, they are grouped based on title and warnings content. The ones from the screenshot have the same title but different details inside.

Makes sense, thanks

### 4
> This part I didn't touch in the PR. I guess the idea was to bring users attention to the fact that they reached the earliest or the latest document and there is nothing else to load. Would you suggest to change this callout?

Can you update following Matthias' suggestion and send a screenshot? So a callout in color primary. We also need a bit of spacing above this banner (so more spacing below the `Load 5 newer documents` line).

jughosta commented:
@andreadelrio 

### 2

> Yes. I don't think the border does a lot to improve contrast tbh and no-border is the recommended pattern in EUI. See https://elastic.github.io/eui/#/display/empty-prompt/guidelines

Thanks!

From eui example:
<img width="400" alt="Screenshot 2023-08-03 at 13 50 08" src="https://github.com/elastic/kibana/assets/1415710/fc2f383c-1551-4501-b147-8d216edb476f">

Based on these guidelines, an icon should be at the top then, no?
<img width="400" alt="Screenshot 2023-08-03 at 13 52 27" src="https://github.com/elastic/kibana/assets/1415710/fc658ff9-256b-41af-b21d-4ce8e8a36159">

Should we also update the icon position for Discover error view? Currently it looks like this:
<img width="400" alt="Screenshot 2023-08-03 at 13 50 16" src="https://github.com/elastic/kibana/assets/1415710/9c4dadad-3781-4588-b407-ee0bb78f10ee">


### 4

> Can you update following Matthias' suggestion and send a screenshot? So a callout in color primary. We also need a bit of spacing above this banner (so more spacing below the Load 5 newer documents line).

Updated the old callout:

<img width="400" alt="Screenshot 2023-08-03 at 11 22 59" src="https://github.com/elastic/kibana/assets/1415710/0d887f7f-54d6-4414-85c8-aeefbd9124d9">

andreadelrio commented:
> From eui example: <img alt="Screenshot 2023-08-03 at 13 50 08" width="400" src="https://user-images.githubusercontent.com/1415710/258103062-fc2f383c-1551-4501-b147-8d216edb476f.png">
> 
> Based on these guidelines, an icon should be at the top then, no? <img alt="Screenshot 2023-08-03 at 13 52 27" width="400" src="https://user-images.githubusercontent.com/1415710/258102917-fc658ff9-256b-41af-b21d-4ce8e8a36159.png">
> 
> Should we also update the icon position for Discover error view? Currently it looks like this: <img alt="Screenshot 2023-08-03 at 13 50 16" width="400" src="https://user-images.githubusercontent.com/1415710/258103383-9c4dadad-3781-4588-b407-ee0bb78f10ee.png">

Yes, I think this makes sense.

davismcphee reviewed (approved):
Tested locally again and the latest changes LGTM 👍 Thanks for updating the error callout too!

kertal commented:
@andreadelrio I've got a theory why the bold was being used, maybe Harry Potter accidentally applied a spell creating a PR to Kibana and used a bolt similar to his scar to claim ownership?
<img width="1317" alt="Bildschirmfoto 2023-08-10 um 11 36 33" src="https://github.com/elastic/kibana/assets/463851/4c0a962a-88c6-4c87-9e5e-53fae2c9480a">
